contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,138 @@ The central manager will poll each worker for their discovered volumes and creat
+if not, add it:
+```
+vgextend exStorageGroup /dev/new
+```
+
+4. Update capacity in StorageClass if there is a change in VG size.
+
+A new field `storageSource` will be added into the configuration ConfigMap, to illustrate available
+storage source for provisioning.
+
+```
+kind: ConfigMap
+metadata:
+  name: local-volume-config
+  namespace: kube-system
+data:
+ storageClassMap: |
+    local-fast:
+      hostDir: "/mnt/ssds"
+    local-slow:
+      hostDir: "/mnt/hdds"
+    local_dynamic:
+      storageSources:
+        - "/dev/sdb"
+        - "/dev/sdc"
+```
+For example for an LVM StorageManager, the list of storageSource could contain disk partitions.
+For other StorageManagers, the list could contain other storage sources (e.g. unpartitioned disks).
+Each StorageManager should know the scope and what to do with the input.
+
+Major downsides of this approach:
+
+* As provisioners running on all nodes shall a same configuration ConfigMap,
+storage sources of their nodes are required to be configured in the same way.
+Take LVM as an example, for configmap above, disks used for provisioning are required
+to be partitioned as "/dev/sdb" and "/dev/sdc".
+
+##### Change in populator
+
+In addition to PV, the `populator` will also be extended to watch PVC objects of all namespaces.
+When it finds:
+
+1. A PVC is newly annotated with `annScheduledNode`, and its node name match the value
+of the annotation
+2. Value of the `annStorageProvisioner` on the PVC match itself
+
+it then trigger volume provsioning, detailed in the volume creation section.
+
+For this purpose, the following permissions are required:
+
+* list/get PersistentVolumeClaims
+
+##### Volume creation
+
+During start-up, a new channel `provisionChan` will be created, and passed into both `populator`
+and `storageManager`:
+```
+provisionChan chan *common.LocalPVConfig
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will consturct an input with necessary config information, and pass it into the channel
+to trigger volume privisioning; `StorageManager` will be the consumer, it will watch the channel and
+manage provisioning when new input received, that is:
+
+1. Create volumes behind a PV accordingly(e.g. exec `lvcreate` for LVM)
+2. Create a PV object that is pre-bound to the PVC
+
+the PV object would be added into existing PV cache too.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+The behavior is extensible enough to support cleanup task of provsioned volumes too.
+So nothing new will be introduced here in the provisioner, but the difference lies
+in the cleanup script/binary itself.
+
+For example, dd-zero could be a way to cleanup static volumes,
+but for dynamically provisioned volumes, `lvremove` is also

```

Comment:
Since the provisioner shall likely create an FS on a create LV, you'll want to `wipefs` before `lvremove`. Rationale: LVM doesn't scrub removed LVs, so if a new one is created immediately after, and `mkfs` is executed, it may fail with `An FS already exists`, and (for safety purposes), I don't think an automated system shall run `mkfs -f` as-is.

Also: how are FS creation options configured?

Reply:
Yes, I did lose a lot of details here.

Initially, I thought the doc is of overall design of local storage, but not limited to LVM, so I mainly described changes of the overall structure to coordinate the API side change, and take LVM things just as an illustration(and get we can detail things about LVMs in another doc, maybe in  repo external-storage) 

Anyway, if you guys agree it should be the place to address LVM details, I'll have an update  to detail the workflow :)

Reply:
> Also: how are FS creation options configured?

fsType and other potential options for FS creation can be specified in StorageClass Parameters,